### PR TITLE
Stop installing as soon as possible after a build failure

### PR DESCRIPTION
**Pull request by dcoutts**
The goal here is to support an install mode where we stop as soon as possible after one package fails, rather than going on as far as possible with other packages.

This is the first patch in a series. More patches will be pushed to this branch. It's here now for early feedback. Subsequent patched will add tests and make use of these new features to stop the install process as soon as possible after a failure, with a new `--keep-going` flag to get the existing behaviour.

First patch: extend the `JobControl` interface
---------------

To properly support this we need to be able to cancel jobs that have been submitted but that have not yet been started. This is easy for the serial JobControl but requires re-implementing the paralle JobControl.

In the previous impl we simply fork a thread immediately for each job (and use a JobLimit externally to limit parallelim). The problem with that is that all submitted jobs are actually executing, so there's no way to cancel jobs that have not started yet (there are none). Combined with the executeInstallPlan behaviour of exposing the maximum possible paralleism this means that there can be a lot of jobs in progress (though most waiting on the JobLimit), and none of them can be sanely cancelled.

So in this impl we have a built-in maximum parallelism, so we're executing at most that many jobs at once, while all others are waiting to be executed. The ones that are waiting can be cancelled.

Because it's unspecified how many jobs can be cancelled, users can no longer rely on counting the number of outstanding jobs so we add a new method to check if there are any jobs remaining.

### CODE REVIEW

File: cabal-install/Distribution/Client/JobControl.hs [hunk 1]
```diff
@@ -27,48 +27,129 @@ module Distribution.Client.JobControl (
   ) where
 
 import Control.Monad
-import Control.Concurrent hiding (QSem, newQSem, waitQSem, signalQSem)
-import Control.Exception (SomeException, bracket_, mask, throw, try)
+import Control.Concurrent (forkIO) --hiding (QSem, newQSem, waitQSem, signalQSem)
```
Comment (dcoutts) [inline cabal-install/Distribution/Client/JobControl.hs]: oops, delete now-unused code...

File: cabal-install/Distribution/Client/JobControl.hs [hunk 2]
```diff
@@ -27,48 +27,129 @@ module Distribution.Client.JobControl (
...
+import Control.Concurrent.MVar
+import Control.Concurrent.STM (STM, atomically)
+import Control.Concurrent.STM.TVar
+import Control.Concurrent.STM.TQueue
```
Comment (dcoutts) [inline cabal-install/Distribution/Client/JobControl.hs]: Note this uses `TQueue` from `stm-2.4`. Are we ok with that version? Alternative could probably use `TChan`.

Comment (23Skidoo) [inline cabal-install/Distribution/Client/JobControl.hs]: @dcoutts I think we are, provided that the bootstrap script will continue to work with older GHCs.

**Comment by dcoutts** [comment]
Did the AppVeyor one fail? There's no indication of failure. Is it a timeout?

**Comment by 23Skidoo** [comment]
Looks like it succeeded.

**Comment by dcoutts** [comment]
@23Skidoo lemme know if you're happy with this. I don't propose to add any more now (assuming the tests pass).

File: cabal-install/Distribution/Client/Update.hs [hunk 3]
```diff
@@ -47,7 +47,7 @@ update verbosity repoCtxt | null (repoContextRepos repoCtxt) = do
   warn verbosity $ "No remote package servers have been specified. Usually "
                 ++ "you would have one specified in the config file."
 update verbosity repoCtxt = do
-  jobCtrl <- newParallelJobControl
+  jobCtrl <- newParallelJobControl 2
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/Update.hs]: Why not make `n` equal to the number of repos?

**Comment by 23Skidoo** [comment]
LGTM modulo one minor comment.

**Comment by dcoutts** [comment]
@23Skidoo ok, will squash and merge when the tests are done.

**Pull request merged**